test/t8n/t8n.cpp

```diff
@@ -0,0 +1,172 @@
+        else if (arg == "--input.alloc" && ++i < argc)
+            alloc_file = argv[i];
+        else if (arg == "--input.env" && ++i < argc)
+            env_file = argv[i];
+        else if (arg == "--input.txs" && ++i < argc)
+            txs_file = argv[i];
+        else if (arg == "--output.basedir" && ++i < argc)
+            output_dir = argv[i];
+        else if (arg == "--output.result" && ++i < argc)
+            output_result_file = argv[i];
+        else if (arg == "--output.alloc" && ++i < argc)
+            output_alloc_file = argv[i];
+    }
+
+    state::BlockInfo block;
+    state::State pre_state;
+
+    try  // FIXME: Remove and use noexcept json::parse function
+    {
+        if (!alloc_file.empty())
+        {
+            const auto j = json::json::parse(std::ifstream{alloc_file}, nullptr, false);
+            pre_state = test::from_json<state::State>(j);
+        }
+        if (!env_file.empty())
+        {
+            const auto j = json::json::parse(std::ifstream{env_file});
+            block = test::from_json<state::BlockInfo>(j);
+        }
+
+        auto state = pre_state;
+
+        std::vector<std::pair<std::string, state::Transaction>> hash_txs;
+
+        json::json j_result;
+        // FIXME: Calculate difficulty properly
+        j_result["currentDifficulty"] = "0x20000";
+        j_result["currentBaseFee"] = tu::shortest_hex(block.base_fee);
+
+        // Parse and execute transactions
+        if (!txs_file.empty())
+        {
+            const auto j_txs = json::json::parse(std::ifstream{txs_file});
+
+            evmc::VM vm{evmc_create_evmone(), {{"O", "0"}}};
+
+            std::vector<state::Log> txs_logs;
+
+            if (j_txs.is_array())
+            {
+                j_result["receipts"] = json::json::array();
+                j_result["rejected"] = json::json::array();
+                auto idx = 0;
+                for (auto& j_tx : j_txs)
+                {
+                    state::Transaction tx = test::from_json<state::Transaction>(j_tx);
+                    auto const [tx_logs, gas_used] = state::transition(state, block, tx, rev, vm);
+                    if (tx_logs.has_value())
+                    {
+                        txs_logs.insert(txs_logs.begin(), (*tx_logs).begin(), (*tx_logs).end());
+                        auto& j_receipt = j_result["receipts"][j_result["receipts"].size()];
+                        j_receipt["transactionHash"] = j_tx["hash"];
+                        j_receipt["gasUsed"] = tu::shortest_hex(uint64_t(gas_used));
+                        j_receipt["cumulativeGasUsed"] = j_receipt["gasUsed"];
+
+                        j_receipt["blockHash"] = NULL_HEXSTRING_32;
+                        j_receipt["contractAddress"] = NULL_HEXSTRING_20;
+                        j_receipt["logsBloom"] = NULL_HEXSTRING_256;
+                        j_receipt["logs"] = json::json::array();  // FIXME: Add to_json<state:Log>
+                        j_receipt["root"] = "";
```

Comment:
Is this really expected to be empty?

Reply:
It's a minimal working version. A lot of fields are required by retesteth but never used.